gnovm/pkg/gnolang/preprocess.go

```diff
@@ -3673,6 +3912,172 @@ func findDependentNames(n Node, dst map[Name]struct{}) {
 	}
 }
 
+func isBlockNodeInLoop(store Store, bn BlockNode) bool {
+	for bn != nil {
+		bn = bn.GetParentNode(store)
+		switch bn.(type) {
+		case *FuncLitExpr:
+			return false
+		case *ForStmt:
+			return true
+		case *RangeStmt:
+			return true
+		default:
+			continue
+		}
+	}
+	return false
+}
+
+// reset value path, static block, preprocessed attr, etc.
+func resetStaticBlock(bn BlockNode) {
+	Transcribe(bn, func(ns []Node, ftype TransField, index int, n Node, stage TransStage) (Node, TransCtrl) {
+		switch stage {
+		case TRANS_ENTER:
+			switch cn := n.(type) {
+			case *NameExpr:
+				cn.Path = NewValuePath(VPUverse, 0, 0, cn.Name)
+				cn.SetAttribute(ATTR_PREPROCESSED, false)
+				return cn, TRANS_CONTINUE
+			case BlockNode:
+				cn.GetStaticBlock().reset()
+				cn.SetAttribute(ATTR_PREPROCESSED, false)
+				return cn, TRANS_CONTINUE
+			default:
+				cn.SetAttribute(ATTR_PREPROCESSED, false)
+				return cn, TRANS_CONTINUE
+			}
+		default:
+			return n, TRANS_CONTINUE
+		}
+	})
+}
+
+// for goto...
+// wrap goto loop into block stmt.
+func rebuildBody(b Body, gloop *LoopInfo, loc Location) Body {
+	preBody := []Stmt{}
+	loopBody := []Stmt{}
+	postBody := []Stmt{}
+
+	for _, s := range b {
+		if s.GetLine() < gloop.labelLine {
+			preBody = append(preBody, s)
+		} else if s.GetLine() >= gloop.labelLine && s.GetLine() <= gloop.gotoLine {
+			loopBody = append(loopBody, s)
+		} else {
+			postBody = append(postBody, s)
+		}
+	}
+
+	nn := BlockS(loopBody)
+	nn.SetLabel(gloop.label) // set label on wrapper block stmt
+
+	// some loc info inherit from parent block
+	nLoc := loc
+	nLoc.Line = loopBody[0].GetLine()
+	nn.SetLocation(nLoc)
+	nn.SetLine(loopBody[0].GetLine())
+
+	nBody := append(preBody, nn)
+	nBody = append(nBody, postBody...)
+	return nBody
+}
+
+// this is only for loops formed by goto...
+func checkAndRebuildBody(body Body, loops []*LoopInfo, loc Location) Body {
+	for _, loop := range loops {
+		if loop.isGotoLoop { // for/range has done the work
+			// find label stmt
+			lblstmt, idx := body.GetLabeledStmt(loop.label)
```

Comment:
i think label conflicts are possible; same label used in different places.

Reply:
This search is conducted within specific block nodes such as funcDecl, forStmt, etc., and not within a fileNode. Thus, each label should be unique within the scope of the search. Did I misunderstand your question?

also, if jump from different places to same label stmt, it's also working,  see this case:

```go
package main

import (
	"fmt"
)

func main() {
	i := 0
	var fs []func()

	defer func() {
		for _, ff := range fs {
			ff()
		}
	}()

	if i == 0 {
		goto Start
	}

Start:
	if i <= 5 {
		x := i
		fmt.Println("i: ", i)
		i++
		fs = append(fs, func() { println(x) })

		goto Start // Jumps back to the label "Start"
	}

	if i <= 10 {
		x := i
		fmt.Println("Counting past 5:", i)
		i++
		fs = append(fs, func() { println(x) })
		goto Start // Another jump back to the label "Start"
	}

	fmt.Println("Finished counting")
}

// Output:
// i:  0
// i:  1
// i:  2
// i:  3
// i:  4
// i:  5
// Counting past 5: 6
// Counting past 5: 7
// Counting past 5: 8
// Counting past 5: 9
// Counting past 5: 10
// Finished counting
// 0
// 1
// 2
// 3
// 4
// 5
// 6
// 7
// 8
// 9
// 10
```